### ISSUE TITLE
Strange failures with Promote/OrdDeriving and Promote/Prelude

### ISSUE THREAD

**Issue by jstolarek** [goldfirere/singletons#118]
Under GHC 7.10 code in these tests promotes without errors but attempting to use promoted definitions fails:

```
Promote/OrdDeriving.hs:0:0:
    Couldn't match type ‘'False’ with ‘'True’
    Expected type: Proxy Bool 'True
      Actual type: Proxy Bool ((:<) Nat 'Zero ('Succ 'Zero))
    In the expression: foo1a
    In an equation for ‘foo1b’: foo1b = foo1a
```

```
Promote/Prelude.hs:0:0:
    Couldn't match type ‘'[] Nat’
                   with ‘(':) Nat 1 ((':) Nat 2 ('[] Nat))’
    Expected type: Proxy [Nat] ((':) Nat 1 ((':) Nat 2 ('[] Nat)))
      Actual type: Proxy
                     [Nat]
                     (Take
                        Nat
                        2
                        ((':) Nat 1 ((':) Nat 2 ((':) Nat 3 ((':) Nat 4 ('[] Nat))))))
    In the expression: takeTest1a
    In an equation for ‘takeTest1b’: takeTest1b = takeTest1a

Promote/Prelude.hs:0:0:
    Couldn't match type ‘(':) Nat 3 ((':) Nat 4 ('[] Nat))’
                   with ‘'[] Nat’
    Expected type: Proxy [Nat] ((':) Nat 3 ((':) Nat 4 ('[] Nat)))
      Actual type: Proxy
                     [Nat]
                     (Drop
                        Nat
                        2
                        ((':) Nat 1 ((':) Nat 2 ((':) Nat 3 ((':) Nat 4 ('[] Nat))))))
    In the expression: dropTest1a
    In an equation for ‘dropTest1b’: dropTest1b = dropTest1a

Promote/Prelude.hs:0:0:
    Couldn't match type ‘(':) Nat 3 ((':) Nat 4 ('[] Nat))’
                   with ‘'[] Nat’
    Expected type: Proxy
                     ((,) [Nat] [Nat])
                     '((':) Nat 1 ((':) Nat 2 ('[] Nat)),
                       (':) Nat 3 ((':) Nat 4 ('[] Nat)))
      Actual type: Proxy
                     ((,) [Nat] [Nat])
                     (SplitAt
                        Nat
                        2
                        ((':) Nat 1 ((':) Nat 2 ((':) Nat 3 ((':) Nat 4 ('[] Nat))))))
    In the expression: splitAtTest1a
    In an equation for ‘splitAtTest1b’: splitAtTest1b = splitAtTest1a
```

Is that our fault or a GHC regression?

**Comment by goldfirere** [comment]
Yikes! This looks bad, and smells faintly of a terrible bug in GHC. Will look into this quite soon.

**Comment by jstolarek** [comment]
I spent some time today trying to reduce this and sadly my reduced example compiles. This either means that I reduced it too much - which might be the case because I really trimmed it a lot - or that we really don't generate all the instances required to reduce these cases.

Here's an observation I made. This fails to compile with the error reported originally in this bug report:
```haskell
foo1a :: Proxy (Zero :< Succ Zero)
foo1a = Proxy

foo1b :: Proxy True
foo1b = foo1a
```
But this works perfectly fine:
```haskell
foox :: Proxy (Succ Zero :< Succ Zero)
foox = Proxy

fooy :: Proxy False
fooy = foox
```
How these two differ? The generated instance of `Compare` does not hit the third case that delegates the call to `Compare` instance for `GHC.TypeLits.Nat`. Perhaps these definitions are not imported into the module? Where are they generated?

**Comment by jstolarek** [comment]
For the record, here's the `Compare` instance derived for our definition of `Nat`:

```haskell
type family Compare_1627988044_a2vAF (a_a2vAG :: Nat_a2vAt)
                                     (b_a2vAH :: Nat_a2vAt) :: Ordering where
  Compare_1627988044_a2vAF Zero_a2vAu Zero_a2vAu = EQ
  Compare_1627988044_a2vAF (Succ_a2vAv x_a2vAM) (Succ_a2vAv y_a2vAN) = ThenCmp EQ (Compare x_a2vAM y_a2vAN)
  Compare_1627988044_a2vAF (a_a2vAG :: Nat_a2vAt) (b_a2vAH :: Nat_a2vAt) = Compare (GetRank_1627988048_a2vAJ a_a2vAG) (GetRank_1627988048_a2vAJ b_a2vAH)
type family GetRank_1627988048_a2vAJ a_a2vAK :: GHC.TypeLits.Nat where
  GetRank_1627988048_a2vAJ Zero_a2vAu = 1
  GetRank_1627988048_a2vAJ (Succ_a2vAv a_a2vAL) = 2
```
The third case requires `Compare` instances for `GHC.TypeLits.Nat`.

**Comment by jstolarek** [comment]
I just tested and the instance `Compare GHC.TypeLits.Nat` is in scope.

I've run out of ideas.

**Comment by goldfirere** [comment]
Took a look at this. It's GHC's fault. But in a usual turn of fate, GHC's bug is my fault, anyway. Filed as GHC#10488, and the fix will be in 7.10.2. Closing this ticket as it doesn't serve much purpose remaining open.

**Comment by peti** [comment]
Is this issue solved in GHC 7.10.2?

**Comment by cstrahan** [comment]
@goldfirere Is this resolved in GHC 7.10.2?

**Comment by goldfirere** [comment]
Yes. This was just a bug in GHC, fixed now, that requires no change in behavior from the singletons library.
